### ISSUE TITLE
`!= NULL` with other condition return the wrong result

### ISSUE THREAD

**Issue by aytrack** [pingcap/tidb#43407]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
CREATE TABLE `UK_HP16600` (
  `COL1` int(11) DEFAULT NULL,
  `COL2` varchar(20) DEFAULT NULL,
  `COL4` datetime DEFAULT NULL,
  `COL3` bigint(20) DEFAULT NULL,
  `COL5` float DEFAULT NULL,
  UNIQUE KEY `UK_COL1` (`COL1`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY HASH (((`COL1`*46))) PARTITIONS 5;
insert into UK_HP16600 values  (1342482627,'阴煥牺鰨瓛瞩垓経繤û紞媪滞镫嘌衍稱貼銂浳','0743-05-03 19:46:22',-235622403343211727,-2.35277e38);
 select  min(col2), max(col2) from UK_HP16600 partition(p2, p3) where col1 != NULL and col1 in (-1268000613, 1342482627, 23627607) group by col1;
```

### 2. What did you expect to see? (Required)
```sql
TiDB root@127.0.0.1:test>  select  min(col2), max(col2) from UK_HP16600 partition(p2, p3) where col1 != NULL and col1 in (-1268000613, 1342482627, 23627607) group by col1;
+-----------+-----------+
| min(col2) | max(col2) |
+-----------+-----------+
+-----------+-----------+
```
### 3. What did you see instead (Required)
```sql
TiDB root@127.0.0.1:test> select  min(col2), max(col2) from UK_HP16600 partition(p2, p3) where col1 != NULL and col1 in (-1268000613, 1342482627, 23627607) group by col1;
+-----------------------------------------+-----------------------------------------+
| min(col2)                               | max(col2)                               |
+-----------------------------------------+-----------------------------------------+
| 阴煥牺鰨瓛瞩垓経繤û紞媪滞镫嘌衍稱貼銂浳 | 阴煥牺鰨瓛瞩垓経繤û紞媪滞镫嘌衍稱貼銂浳 |
+-----------------------------------------+-----------------------------------------+

TiDB root@127.0.0.1:test> explain select  min(col2), max(col2) from UK_HP16600 partition(p2, p3) where col1 != NULL and col1 in (-1268000613, 1342482627, 23627607) group by col1;
+------------------------------+---------+-----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| id                           | estRows | task      | access object                         | operator info                                                                                                       |
+------------------------------+---------+-----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| HashAgg_8                    | 1.00    | root      |                                       | group by:test.uk_hp16600.col1, funcs:min(test.uk_hp16600.col2)->Column#7, funcs:max(test.uk_hp16600.col2)->Column#8 |
| └─IndexLookUp_17             | 1.00    | root      | partition:p2,p3                       |                                                                                                                     |
|   ├─IndexRangeScan_15(Build) | 1.00    | cop[tikv] | table:UK_HP16600, index:UK_COL1(COL1) | range:[-1268000613,-1268000613], [23627607,23627607], [1342482627,1342482627], keep order:false                     |
|   └─TableRowIDScan_16(Probe) | 1.00    | cop[tikv] | table:UK_HP16600                      | keep order:false                                                                                                    |
+------------------------------+---------+-----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------+
4 rows in set
```
### 4. What is your TiDB version? (Required)
```sql
TiDB root@127.0.0.1:test> select tidb_version();
+-----------------------------------------------------------+
| tidb_version()                                            |
+-----------------------------------------------------------+
| Release Version: v7.1.0-rc.0                              |
| Edition: Community                                        |
| Git Commit Hash: ba71c5fcedf6da2ec68759513493b7484e2e97b5 |
| Git Branch: heads/refs/tags/v7.1.0-rc.0                   |
| UTC Build Time: 2023-04-22 10:11:11                       |
| GoVersion: go1.20.3                                       |
| Race Enabled: false                                       |
| TiKV Min Version: 6.2.0-alpha                             |
| Check Table Before Drop: false                            |
| Store: tikv                                               |
+-----------------------------------------------------------+
```

**Comment by aytrack** [comment]
since it's  a regression, upgrade to critical

**Comment by chrysan** [comment]
It's related with a new rule predicate_simplification @ghazalfamilyusa. Looks like we didn't cover tests for partition table + predicate_simplification. If I remove the new rule, the correct plan is:
```
+----------------------------------+---------+-----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| id                               | estRows | task      | access object                         | operator info                                                                                                       |
+----------------------------------+---------+-----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| HashAgg_8                        | 1.00    | root      |                                       | group by:test.uk_hp16600.col1, funcs:min(test.uk_hp16600.col2)->Column#7, funcs:max(test.uk_hp16600.col2)->Column#8 |
| └─IndexLookUp_19                 | 1.00    | root      | partition:p2,p3                       |                                                                                                                     |
|   ├─Selection_18(Build)          | 1.00    | cop[tikv] |                                       | ne(test.uk_hp16600.col1, NULL)                                                                                      |
|   │ └─IndexRangeScan_16          | 1.00    | cop[tikv] | table:UK_HP16600, index:UK_COL1(COL1) | range:[-1268000613,-1268000613], [23627607,23627607], [1342482627,1342482627], keep order:false                     |
|   └─TableRowIDScan_17(Probe)     | 1.00    | cop[tikv] | table:UK_HP16600                      | keep order:false                                                                                                    |
+----------------------------------+---------+-----------+---------------------------------------+---------------------------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```
There's a selection.

**Comment by ghazalfamilyusa** [comment]
The issue is not specific to partitioned  tables and root cause is that that not equality with NULL has different semantics in mysql than general non equality. x != NULL is always false while x != other-constant could be true or false. A smaller test case is below where the result should be empty and instead we remove a1 != NULL against a1 in (1,2,3) which is not correct.

create table t1 (a1 int, b1 int)
insert into t1 values (1,1)

select * from t1 where a1 != null and a1 in (1,2,3)
--------------
a1      b1
1       1

explain select * from t1 where a1 != null and a1 in (1,2,3)
--------------
id      estRows task    access object   operator info
TableReader_7   30.00   root            data:Selection_6
└─Selection_6   30.00   cop[tikv]               **in(test.t1.a1, 1, 2, 3)**
  └─TableFullScan_5     10000.00        cop[tikv]       table:t1        keep order:false, stats:pseudo

**Comment by time-and-fate** [comment]
closed by #43434
